### PR TITLE
MAKE-1213: Restrict calculation of change points to the portion of the time series after the latest triaged change point

### PR DESCRIPTION
**Pull request by ancostas**
Sorry this got a bit involved on the testing side; I felt our current test was a bit too rigid (is this passing because it's a super simple case?) and needed to be a bit more dynamic (mimicking what cedar actually will encounter).

Now it:
* Generates a random set of time series, of random lengths
* Generates a random number of change points in these time series
* Feeds them in randomly grouped under perf results (rollups can be added/removed over time)
* Ensures that they all get detected and persisted as you would expect them to be

Also, I added a test the restricted analysis this ticket is actually about :smile:

### CODE REVIEW

**Comment by ancostas** [comment]
Not sure why test-units is failing... looks like an off by one error I handled early in setting this up, and this passes for me locally :raised_eyebrow:  -- will look into it tomorrow; feel free to make the initial pass in the meantime.

**Comment by ancostas** [comment]
@tychoish Idk what's wrong here; this works for me locally. It's like it's not able to get results using the same aggregation when run through evergreen?

File: model/change_point_detection.go [hunk 1]
```diff
@@ -428,3 +461,18 @@ func createChangePoint(ctx context.Context, env cedar.Environment, resultToUpdat
...
+		perfIDKey: perfResultId,
+		bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, perfChangePointMeasurementKey): measurement,
+	}
+	update := bson.M{
+		"$set": bson.M{
+			bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, "$", perfChangePointTriageKey, perfTriageInfoStatusKey):    status,
+			bsonutil.GetDottedKeyName(perfAnalysisKey, perfAnalysisChangePointsKey, "$", perfChangePointTriageKey, perfTriageInfoTriagedOnKey): time.Now(),
+		},
+	}
+	_, err := env.GetDB().Collection(perfResultCollection).UpdateOne(ctx, filter, update)
```
Comment (tychoish) [inline model/change_point_detection.go:476]: you have to check the result (first returned value). the error value is only non-nil when there's a connection, parsing, or other related error, if you failed to find or update a document, that's in the result.

File: model/change_point_detection.go [hunk 2]
```diff
@@ -428,3 +461,18 @@ func createChangePoint(ctx context.Context, env cedar.Environment, resultToUpdat
 	_, err := env.GetDB().Collection(perfResultCollection).UpdateOne(ctx, filter, update)
 	return errors.Wrap(err, "Unable to create change point")
 }
+
+func TriageChangePoint(ctx context.Context, env cedar.Environment, perfResultId string, measurement string, status TriageStatus) error {
```
Comment (tychoish) [inline model/change_point_detection.go:465]: I'd be consistent about using ID rather than Id

**Comment by tychoish** [comment]
I'm also worried about the failing test, or if we need to be careful or should expect anything unusual in the deploy.

File: model/change_point_detection.go (hunk 2 above)
Comment (ancostas) [reply model/change_point_detection.go:465]: Sorry, force of habit!

**Comment by ancostas** [comment]
@tychoish Tests were failing since we use 4.0.13 when testing through evergreen, and `$set` operator in aggregations isn't supported until 4.2. This was... pretty troublesome to debug.

File: units/recalculate_change_points_test.go [hunk 3]
```diff
@@ -10,21 +12,95 @@ import (
...
+				break
+			}
+		}
+	}
+	return newVals
+}
+
+func makePerfResultsWithChangePoints(unique string, seed int64) ([]testResultsAndRollups, map[string][]int) {
+	// deterministic testing on failure
+	fmt.Println("Seed for recalculate test: " + strconv.FormatInt(seed, 10))
```
Comment (tychoish) [inline units/recalculate_change_points_test.go:46]: can you omit printing in tests (use the logger, if you must.)

File: units/recalculate_change_points_test.go [hunk 4]
```diff
@@ -34,33 +110,35 @@ func makePerfResults(length int, breakpoint int, unique string) []TestResultsAnd
...
+	for measurement, cpIndices := range timeSeriesChangePoints {
+		measurementName := "measurement_" + strconv.Itoa(measurement)
+		changePoints[measurementName] = append(changePoints[measurementName], cpIndices...)
+	}
+
+	return rollups, changePoints
 }
 
-func init() {
+func setup() {
```
Comment (tychoish) [inline units/recalculate_change_points_test.go:141]: nit: give this a less generic name

**Pull request merged**